### ISSUE TITLE
Device::create_texture panics in a specific context

### ISSUE THREAD

**Issue by thenlevy** [gfx-rs/wgpu#1165]
**Description**
This is an error that I have faced only on one specific windows machine.
The examples on wgpu-rs work fine but my program hosted on https://github.com/thenlevy/icednano panics when creating a texture with the following error:
```
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `0`,
 right: `-2005270523`', C:\Users\nlevy\.cargo\registry\src\github.com-1ecc6299db9ec823\gfx-backend-dx12-0.6.13\src\device.rs:2634:9
```

**Repro steps**
Running `cargo run` at the root of https://github.com/thenlevy/icednano
this does reproduce the issue on only one machine that I have worked with. I have run the program successfully on other windows machine

**Extra materials**

Here is the output of `Adapter::get_info()`: 
```
AdapterInfo { name: "Microsoft Basic Render Driver", vendor: 5140, device: 140, device_type: IntegratedGpu, backend: Dx12 }
```

Here is the descriptor of the texture I am trying to create: `TextureDescriptor { label: Some("diffuse_texture"), size: Extent3d { width: 512, height: 512, depth: 1 }, mip_level_count: 9, sample_count: 1, dimension: D2, format: Rgba8UnormSrgb, usage: COPY_DST | SAMPLED }`

Here is a full backtrace
```
stack backtrace:
   0:     0x7ff6340515ae - std::backtrace_rs::backtrace::dbghelp::trace
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\..\..\backtrace\src\backtrace\dbghelp.rs:108
   1:     0x7ff6340515ae - std::backtrace_rs::backtrace::trace_unsynchronized
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\..\..\backtrace\src\backtrace\mod.rs:66
   2:     0x7ff6340515ae - std::sys_common::backtrace::_print_fmt
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\sys_common\backtrace.rs:67
   3:     0x7ff6340515ae - std::sys_common::backtrace::_print::{{impl}}::fmt
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\sys_common\backtrace.rs:46
   4:     0x7ff6340682cb - core::fmt::write
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\core\src\fmt\mod.rs:1078
   5:     0x7ff63404dc88 - std::io::Write::write_fmt<std::sys::windows::stdio::Stderr>
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\io\mod.rs:1518
   6:     0x7ff63405439d - std::sys_common::backtrace::_print
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\sys_common\backtrace.rs:49
   7:     0x7ff63405439d - std::sys_common::backtrace::print
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\sys_common\backtrace.rs:36
   8:     0x7ff63405439d - std::panicking::default_hook::{{closure}}
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\panicking.rs:208
   9:     0x7ff634053f78 - std::panicking::default_hook
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\panicking.rs:227
  10:     0x7ff634054d0f - std::panicking::rust_panic_with_hook
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\panicking.rs:593
  11:     0x7ff634054811 - std::panicking::begin_panic_handler::{{closure}}
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\panicking.rs:499
  12:     0x7ff634051eef - std::sys_common::backtrace::__rust_end_short_backtrace<closure-0,!>
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\sys_common\backtrace.rs:141
  13:     0x7ff634054769 - std::panicking::begin_panic_handler
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\panicking.rs:495
  14:     0x7ff63405471c - std::panicking::begin_panic_fmt
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\panicking.rs:437
  15:     0x7ff633fdf5fa - gfx_backend_dx12::device::<impl gfx_hal::device::Device<gfx_backend_dx12::Backend> for gfx_backend_dx12::Device>::bind_image_memory::hbf2569b0d267b02b
  16:     0x7ff633ea4013 - wgpu_core::device::Device<B>::create_texture::h6185c1f03652bf85
  17:     0x7ff633de77f0 - wgpu_core::device::<impl wgpu_core::hub::Global<G>>::device_create_texture::hd9dd72a0d32e5a53
  18:     0x7ff633d361dd - <wgpu::backend::direct::Context as wgpu::Context>::device_create_texture::hb40aad04d45025a2
  19:     0x7ff633d092c4 - wgpu::Device::create_texture::hb2be35e7ed8f3d7d
  20:     0x7ff63387cf76 - icednano::text::Letter::new::h1bb5fb0a59ba7659
  21:     0x7ff63391fe14 - <core::iter::adapters::Map<I,F> as core::iter::traits::iterator::Iterator>::fold::habb96be7dab8e9da
  22:     0x7ff6337f5b34 - icednano::scene::Scene::new::h5f6d16cd8fbafcd9
  23:     0x7ff63387f246 - icednano::text::Letter::new::h1bb5fb0a59ba7659
  24:     0x7ff6339c01d6 - std::sys_common::backtrace::__rust_begin_short_backtrace::h1c3f06e2b11a7d7d
  25:     0x7ff63396e46c - std::rt::lang_start::{{closure}}::hae84a9d3425ac407
  26:     0x7ff634055053 - core::ops::function::impls::{{impl}}::call_once
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\library\core\src\ops\function.rs:280
  27:     0x7ff634055053 - std::panicking::try::do_call
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\panicking.rs:381
  28:     0x7ff634055053 - std::panicking::try
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\panicking.rs:345
  29:     0x7ff634055053 - std::panic::catch_unwind
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\panic.rs:396
  30:     0x7ff634055053 - std::rt::lang_start_internal
                               at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca\/library\std\src\rt.rs:51
  31:     0x7ff633884e47 - main
  32:     0x7ff634124b40 - invoke_main
                               at d:\agent\_work\3\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:78
  33:     0x7ff634124b40 - __scrt_common_main_seh
                               at d:\agent\_work\3\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
  34:     0x7ff98c377974 - BaseThreadInitThunk
  35:     0x7ff98c6ea0b1 - RtlUserThreadStart
```

**Comment by kvark** [comment]
Thank you for providing all the details about the issue!

Oh fun, you are running on WARP ("Microsoft Basic Render Driver"). This may happen if your HW is really old, it doesn't support either Vulkan, or DX12. That is to say, we fully expect to work properly on WARP as well, so this is still a bug. It's just that at the end of the day, you'll probably want to use DX11 on that machine as well instead of WARP, if/when DX11 backend is good enough. I wonder if wgpu will pick it for you automatically if you `BackendBit` mask includes DX11?

The next step in this issue would be running within Visual Studio. It would spew DX runtime messages in the output window, and there it would tell us what exactly went wrong. If you can provide that output, that would be great!

**Comment by thenlevy** [comment]
Interesting result, when using the DX11 backend, the program panics when run with `cargo run --release` in a console but when the executable in target/release is run in visual studio, the program doesn't crash immediately but the textures are weird, and the program crashes after some user interaction, this times the panicking function is `Device::create_buffer`

Here are the adapter info, the visual studio messages and the backtraces for various backend bits:

* `BackendBit::Primary`
```
AdapterInfo { name: "Microsoft Basic Render Driver", vendor: 5140, device: 140, device_type: IntegratedGpu, backend: Dx12 }
```
[primary_visual_output.txt](https://github.com/gfx-rs/wgpu/files/5862980/visual_output.txt)
[full_backtrace.txt](https://github.com/gfx-rs/wgpu/files/5862996/full_backtrace.txt)





* `BackendBit::Primary | BackendBit::DX11`
```
AdapterInfo { name: "Microsoft Basic Render Driver", vendor: 5140, device: 140, device_type: IntegratedGpu, backend: Dx12 }
```
(the adapter as with primary only)
[visual_output.txt](https://github.com/gfx-rs/wgpu/files/5862983/visual_output.txt)
[both_full_backtrace.txt](https://github.com/gfx-rs/wgpu/files/5862994/full_backtrace.txt)


* `BackendBit::DX11`
```
AdapterInfo { name: "Microsoft Basic Render Driver", vendor: 5140, device: 140, device_type: IntegratedGpu, backend: Dx11 }
```
[dx11_visual_output_no_crash.txt](https://github.com/gfx-rs/wgpu/files/5862987/visual_output_no_crash.txt)
With Dx11 backend, the program does not crash immediately but crashes after some user interaction, the faulty function being `Device::create_buffer`
[dx11_full_backtrace.txt](https://github.com/gfx-rs/wgpu/files/5862993/full_backtrace.txt)

**Comment by kvark** [comment]
Was able to reproduce on Windows, and surprisingly DX runtime *did* print the exact offender.
The problem is - you have a buffer of size = 0, and you are creating a binding from it. This is not valid in WebGPU.

**Comment by kvark** [comment]
Fix on master is coming in #1173. I don't think we want to back-port it, as we are trying to release 0.7 soon.
The fix on your side would be - not using bindings of size 0 (or buffers of size 0).
